### PR TITLE
VM: Strip zeros when putting contract storage in StateManager

### LINKED ISSUE #875: Check situation if a zero-Buffer is put into contract storage
Related: #729 

It is/was possible to put zero-Buffers (i.e. `Buffer.from('00', 'hex')`) in StateManager via `putContractStorage`. This messes up state trie calculations. It looks like in the MPT itself this is [fixed](https://github.com/ethereumjs/merkle-patricia-tree/blob/aa9e47cf2e9f5d35bd1978e69b2c39046f05b6ff/src/secure.ts#L37), but it might be that this is not fixed in the VM due to our cache. (Placeholder issue)

### PR DESCRIPTION
**Pull request by jochem-brouwer**
Closes #875 

Had a bit of a wrong understanding here, storage slot values should be stripped of leading zeros, which is now enforced in `putContractStorage`. Since putting >32 bytes storage values is not possible in EVM, this is now also disallowed. If we strip leading zeros this means that any zero-byte will immediately be detected as "delete value", both in StateManager but also in MPT, which is nice.

### CODE REVIEW

**Comment by jochem-brouwer** [comment]
Right, tests are failing, but reading the errors seems like this is a case of "the tests are wrong". We might want to consider switching the tests to TypeScript at some point too, because looks like numbers get fed into `putContractStorage` where we expect Buffers. Will fix.

**Comment by holgerd77** [comment]
@jochem-brouwer yes, that was exactly what I was thinking when reviewing #849 (switching to TypeScript), probably rather sooner than later since this should give direct benefits on reliability.

//cc @ryanio

File: packages/vm/lib/state/stateManager.ts [hunk 1]
```diff
@@ -276,13 +276,19 @@ export default class DefaultStateManager implements StateManager {
...
   async putContractStorage(address: Buffer, key: Buffer, value: Buffer): Promise<void> {
     if (key.length !== 32) {
       throw new Error('Storage key must be 32 bytes long')
     }
 
+    if (value.length > 32) {
+      throw new Error('Storage key cannot be longer than 32 bytes')
+    }
+
+    value = unpadBuffer(value)
```
Comment (holgerd77) [inline packages/vm/lib/state/stateManager.ts:290]: You are writing in the comments that a value will be left-padded and then you actually unpad (I read the semantics of this function as: "remove the padding", correct me if I am wrong) here, how does this go together?

Comment (jochem-brouwer) [reply packages/vm/lib/state/stateManager.ts:290]: Oops this is a comment from when I was still thinking that storage slots should always be 32 bytes; this is thus not the case, the leading zeros are stripped. Will update

File: packages/vm/tests/api/istanbul/eip-2200.js [hunk 2]
```diff
@@ -46,7 +46,7 @@ tape('Istanbul: EIP-2200: net-metering SSTORE', async t => {
     await vm.stateManager.putAccount(addr, account)
     await vm.stateManager.putContractCode(addr, Buffer.from(testCase.code, 'hex'))
     if (!testCase.original.isZero()) {
-      await vm.stateManager.putContractStorage(addr, key, testCase.original)
+      await vm.stateManager.putContractStorage(addr, key, testCase.original.toBuffer())
```
Comment (jochem-brouwer) [inline packages/vm/tests/api/istanbul/eip-2200.js:49]: @evertonfraga The browser tests on this PR are hanging, this might be the culprit?

Comment (jochem-brouwer) [reply packages/vm/tests/api/istanbul/eip-2200.js]: Seems to be

File: packages/vm/lib/state/stateManager.ts [hunk 3]
```diff
@@ -415,7 +421,7 @@ export default class DefaultStateManager implements StateManager {
   }
 
   /**
-   * Dumps the the storage values for an `account` specified by `address`.
+   * Dumps the the RLP-encoded storage values for an `account` specified by `address`.
```
Comment (ryanio) [inline packages/vm/lib/state/stateManager.ts:424]:
```suggestion
   * Dumps the RLP-encoded storage values for an `account` specified by `address`.
```

File: packages/vm/lib/state/stateManager.ts [hunk 4]
```diff
@@ -276,13 +276,19 @@ export default class DefaultStateManager implements StateManager {
    * corresponding to `address` at the provided `key`.
    * @param address -  Address to set a storage value for
    * @param key - Key to set the value at. Must be 32 bytes long.
-   * @param value - Value to set at `key` for account corresponding to `address`
+   * @param value - Value to set at `key` for account corresponding to `address`. Cannot be more than 32 bytes. Leading zeros are stripped. If it is a empty or filled with zeros, delete the value.
```
Comment (ryanio) [inline packages/vm/lib/state/stateManager.ts:279]:
```suggestion
   * @param value - Value to set at `key` for account corresponding to `address`. Cannot be more than 32 bytes. Leading zeros are stripped. If it is a empty or filled with zeros, deletes the value.
```

**Comment by ryanio** [comment]
this LGTM! nice additional tests.

it just needs an update to ts now, then can approve and merge.

**Comment by jochem-brouwer** [comment]
Rebased and updated the spelling/grammar errors in the docs, should be ready @ryanio 😄

**Comment by jochem-brouwer** [comment]
Ping for re review @holgerd77 @ryanio 😄

**Comment by holgerd77** [comment]
@jochem-brouwer are you doing this "right now"? Otherwise should be also ok to use the "Update branch" function (rebase is a bit better of course) and I would do a review during the next hour or so.

**Comment by jochem-brouwer** [comment]
Rebasing it now!

**Comment by jochem-brouwer** [comment]
This was an easy rebase, no conflicts 😄  Ready for review @holgerd77

File: packages/vm/lib/state/stateManager.ts [hunk 5]
```diff
@@ -276,13 +276,19 @@ export default class DefaultStateManager implements StateManager {
...
-   * @param value - Value to set at `key` for account corresponding to `address`
+   * @param value - Value to set at `key` for account corresponding to `address`. Cannot be more than 32 bytes. Leading zeros are stripped. If it is a empty or filled with zeros, deletes the value.
    */
   async putContractStorage(address: Buffer, key: Buffer, value: Buffer): Promise<void> {
     if (key.length !== 32) {
       throw new Error('Storage key must be 32 bytes long')
     }
 
+    if (value.length > 32) {
+      throw new Error('Storage key cannot be longer than 32 bytes')
```
Comment (holgerd77) [inline packages/vm/lib/state/stateManager.ts:287]: "key" -> "value"

File: packages/vm/tests/api/state/stateManager.spec.ts [hunk 6]
```diff
@@ -409,3 +409,80 @@ tape('StateManager - Contract code', (tester) => {
...
+    try {
+      await stateManager.putContractStorage(address, key, value)
+      t.fail('did not throw')
+    } catch (e) {
+      t.pass('threw on trying to set storage values larger than 32 bytes')
+    }
+    t.end()
+  })
+
+  it('should strip storage values lower than 32 bytes', async (t) => {
```
Comment (holgerd77) [inline packages/vm/tests/api/state/stateManager.spec.ts:431]: This test description is a bit misleading (stripping is logically not connected to the byte length) respectively also not correct (stripping is also done for 32-byte length. Can you please adjust?

File: packages/vm/tests/api/state/stateManager.spec.ts [hunk 7]
```diff
@@ -409,3 +409,80 @@ tape('StateManager - Contract code', (tester) => {
...
+    const stateManager = new DefaultStateManager()
+    const address = zeros(20)
+    const key0 = zeros(32)
+    const value0 = Buffer.from('aa'.repeat(31), 'hex') // put a value of 31-bytes length
+    const expect0 = unpadBuffer(value0)
+    const key1 = Buffer.concat([zeros(31), Buffer.from('01', 'hex')])
+    const value1 = Buffer.from('aa'.repeat(1), 'hex') // put a vlaue of 1-byte length
```
Comment (holgerd77) [inline packages/vm/tests/api/state/stateManager.spec.ts:438]: typo in comment

File: packages/vm/tests/api/state/stateManager.spec.ts [hunk 8]
```diff
@@ -409,3 +409,80 @@ tape('StateManager - Contract code', (tester) => {
...
+    const expect1 = unpadBuffer(value1)
+
+    await stateManager.putContractStorage(address, key0, value0)
+    const slot0 = await stateManager.getContractStorage(address, key0)
+    t.ok(slot0.equals(expect0), 'value of 31 bytes padded correctly')
```
Comment (holgerd77) [inline packages/vm/tests/api/state/stateManager.spec.ts:443]: Really nit, but I would think this test would be a bit better readable if the respective execution part of these two cases are combined with the value setup (so everything with *0 together and everything with *1 together)

**Review by holgerd77 — COMMENTED** [review]
Some change requests on the comments, otherwise looks good. 😄

File: packages/vm/tests/api/state/stateManager.spec.ts (hunk 8 above)
Comment (jochem-brouwer) [reply packages/vm/tests/api/state/stateManager.spec.ts:440]: Fixed -> 5217f50

Note: I have edited this test a little bit (it did `unpadZeros` in previous commits on this particular test, but none of the test cases actually had leading zeros)

File: packages/vm/tests/api/state/stateManager.spec.ts (hunk 7 above)
Comment (jochem-brouwer) [reply packages/vm/tests/api/state/stateManager.spec.ts]: Fixed in 5217f50

File: packages/vm/tests/api/state/stateManager.spec.ts (hunk 6 above)
Comment (jochem-brouwer) [reply packages/vm/tests/api/state/stateManager.spec.ts]: I have edited this name in 5217f50, hope this is better. I think this name was from when I still expected storage values to always be 32 bytes and I did not edit it right when I realized that was wrong.

File: packages/vm/lib/state/stateManager.ts (hunk 5 above)
Comment (jochem-brouwer) [reply packages/vm/lib/state/stateManager.ts]: Fixed in 5217f50

**Comment by jochem-brouwer** [comment]
Thanks for the review @holgerd77, had no time this afternoon to take a further look but have fixed (?) it now 😄

**Review by holgerd77 — APPROVED** [review]
Ok, this looks good now, thanks @jochem-brouwer, great find! Sorry that this took a bit long, will merge now.

**Pull request merged**